### PR TITLE
Feature/milestone friendship v1

### PR DESCRIPTION
**Pull request by jpdf00**
This milestone adds friendship functionality:
- Users are able to send friendship invitations to other users (but only once, and not to themselves).
- Users see a button “Invite to friendship” next to the name of the user who is not their friend yet - on both users’ list and single-user page.
- Users are able to see pending friendship invitations sent to them from other users.
- Users are able to accept or reject friendship invitations sent to them.

### CODE REVIEW

**Review by GzimAsani — CHANGES_REQUESTED** [review]
# Status: Changes Required :red_circle: 

### Hello **team**, first of all I apologize for the time you waited for this review. I had difficulties running your migrations and this was my problem. Unfortunately I was not able to review your app completely I can only say that user/friendship model is looking  good and so far I can only suggest you to do the following:
- [x] Add  Unit and Integration tests!  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/tests.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
- [x] There should not be any logic in the views because the latter is used only for rendering. Please move all your logic from your views and add it in your helpers.
- [x] You have rubocop offenses locally, kindly fix them

Please do those required changes and ask for another code review where the next reviewer can check your app layouts and its functionality.

_**If you have any question related to the required changes feel free to text me on slack @GzimAsani** or just add a comment here_

HAPPY CODING :computer: :desktop_computer:

File: spec/integration_spec.rb [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+#     fill_in 'Email', with: user.email
+#     fill_in 'Password', with: user.password
+#     click_on 'Log in'
+#     sleep(3)
+#     click_on 'Sign Out'
+#     sleep(3)
+#     visit root_path
+#     expect(page).to have_content('Sign In')
+#   end
+# end
```
Comment (acushlakoncept) [inline spec/integration_spec.rb]: It’s better to delete **commented piece of codes**. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))
See this [post](https://stackoverflow.com/questions/13130859/inline-comments-in-ruby/13130900#:~:text=No%2C%20Ruby%20does%20not%20have,comment%20out%20the%20one%20row), about why comments are discouraged.

File: spec/models_spec.rb [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+      aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+      aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+      aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+      aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+      aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+      aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+      aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+      aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'
+      expect(post).to_not be_valid
```
Comment (acushlakoncept) [inline spec/models_spec.rb]: You can save some lines by simple using `times` method to generate your 1000 character string
```ruby
str = ' '
1000.times { str << 'a' }
post.content = str
```

File: app/helpers/user_helper.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+module UserHelper
+  def added_friend(user, friendship)
+    if !current_user.id == user.id && !Friendship.where(user_id: current_user.id,
+                                                        friend_id: user.id).or(Friendship.where(
+                                                                                 friend_id: current_user.id,
+                                                                                 user_id: user.id
+                                                                               )).exists?
```
Comment (acushlakoncept) [inline app/helpers/user_helper.rb]: you should never use the `where` method, or any other query-building methods like this, outside of the model class itself, Take a look at this article for more Ruby on Rails best practices: [Buggy Rails Code: The 10 Most Common Mistakes That Rails Developers Make](https://www.toptal.com/ruby-on-rails/top-10-mistakes-that-rails-programmers-make)

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hello @jpdf00  and @anapdh, 

You've done well so far 👍 and I commend you for that 👏
However, there are still some changes to be made before this milestone can be approved.

- [x] @anapdh your commit messages should be capitalized. You can watch this [video](https://www.loom.com/share/00771f369b7b47fc968c019ecf0d3ba6) on how to edit your commit messages using git rebase
 See [Git Commit Best Practices](https://github.com/trein/dev-best-practices/wiki/Git-Commit-Best-Practices#formatting-rules) for more insights

- [x] User should be able to invite friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/send_invitation.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

* [x] User should be able to see friendships requests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

* [x] User has to accept friendship request to start friendship ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

* [x] User can reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

* [x] There should be a button to invite user to friendship on users list page and on single user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

* [x] Posts timeline should include posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

- [x] It's not really a good idea boxing all your unit tests into one single file. Let's keep organize and professional. You can create a model folder inside the spec directory, and let each model have their on spec file. Same for integration tests.

- [x] There should be NO application logic in the views nor helpers (no Post.new, Comment.create, current_user.friends.find_by(...)...). Everything connected with data processing should be encapsulated in models and used in controllers. I have made some comments on the affected method in your codebase below.

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla) or reply to this thread, if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi Team     
#### Great Job so far, everything works fine, but there are a few more things required. 
### General

- [x] This should have its own page, assuming I had a thousand posts I have to scroll all the way down 
Screenshot from 2021-03-16 23-56-51
- [x] Kindly separate your test into files it looks more organised and professional.

### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Review by NtwaliHeritier — APPROVED** [review]
### **STATUS: APPROVED✔✔✔✔**

Great  implementation of the project. 
You can merge👏👏👏

Happy coding💻💻

**Pull request merged**